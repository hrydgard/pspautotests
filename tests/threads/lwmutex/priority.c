#include "shared.h"

SceLwMutexWorkarea workarea;

static int threadFunction(SceSize argSize, void* argPointer) {
	int num = argPointer ? *((int*)argPointer) : 0;
	schedf("A%d\n", num);
	sceKernelLockLwMutexCB(&workarea, 1, NULL);
	schedf("B%d\n", num);
	sceKernelUnlockLwMutex(&workarea, 1);

	return 0;
}

void execPriorityTests(int attr, int deleteInstead) {
	SceUID threads[5];
	int test[5] = {1, 2, 3, 4, 5};
	int result;

	sceKernelCreateLwMutex(&workarea, "mutex1", attr, 1, NULL);

	int i;
	for (i = 0; i < 5; i++) {
		threads[i] = CREATE_PRIORITY_THREAD(threadFunction, 0x16 - i);
		sceKernelStartThread(threads[i], sizeof(int), (void*)&test[i]);
	}

	// This one intentionally is an invalid unlock.
	sceKernelDelayThread(1000);
	schedf("Unlocking...\n");
	result = sceKernelUnlockLwMutex(&workarea, 2);
	sceKernelDelayThread(5000);
	schedf("Unlocked 2? %08X\n", result);

	if (!deleteInstead)
	{
		sceKernelDelayThread(1000);
		schedf("Unlocking...\n");
		result = sceKernelUnlockLwMutex(&workarea, 1);
		sceKernelDelayThread(5000);
		schedf("Unlocked 1? %08X\n", result);
	}

	sceKernelDelayThread(1000);
	schedf("Delete: %08X\n", sceKernelDeleteLwMutex(&workarea));
	schedf("\n\n");
	flushschedf();
}

int main(int argc, char **argv) {
	execPriorityTests(0x000, 0);
	execPriorityTests(0x100, 0);
	execPriorityTests(0x000, 1);
	execPriorityTests(0x100, 1);
	return 0;
}