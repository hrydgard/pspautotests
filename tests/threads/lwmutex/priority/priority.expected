A1
A2
A3
A4
A5
Unlocking...
Unlocked 2? 800201BD
Unlocking...
B1
B2
B3
B4
B5
Unlocked 1? 00000000
Delete: 00000000


A1
A2
A3
A4
A5
Unlocking...
Unlocked 2? 800201BD
Unlocking...
B5
B4
B3
B2
B1
Unlocked 1? 00000000
Delete: 00000000


A1
A2
A3
A4
A5
Unlocking...
Unlocked 2? 800201BD
B5
B4
B3
B2
B1
Delete: 00000000


A1
A2
A3
A4
A5
Unlocking...
Unlocked 2? 800201BD
B5
B4
B3
B2
B1
Delete: 00000000


